packages/gatsby/src/redux/nodes.ts

```diff
@@ -419,37 +419,25 @@ function addNodeToBucketWithElemMatch(
 const binarySearch = (
   values: Array<FilterValue>,
   needle: FilterValue
-): [number, number] | undefined => {
+): number => {
   let min = 0
   let max = values.length - 1
-  let pivot = Math.floor(values.length / 2)
   while (min <= max) {
+    const pivot = min + Math.floor((max - min)/ 2)
     const value = values[pivot]
     if (needle < value) {
       // Move pivot to middle of nodes left of current pivot
       // assert pivot < max
-      max = pivot
+      max = pivot - 1
```

Comment:
I've restructured things a bit after your feedback, but this feature remains.

The basic answer is: yes. If it does, then the while loop will exit on the next iteration, since it necessarily means `min > max` (`min` is always `>= 0` and max is always `<= values.length - 1`). This is totally safe. If `max==-1` at the end of the loop, then `pivot==0` at the top of the loop, which means `min==0` and `max==1 || max==0`. In either case, `min` remains `0` and the index we're looking for is `0`. In the other case, if `min > values.length - 1` at the end of the loop, then `pivot == values.length - 1` at the top of loop and the predicate is still `true`, so the index we're looking for is actually `values.length`.